study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/task/TestBostonAddSampleCollectMessage.java

```diff
@@ -0,0 +1,105 @@
+
+    private Config varsCfg;
+    private Config dataCfg;
+    private Gson gson;
+
+    @Override
+    public void init(Path cfgPath, Config studyCfg, Config varsCfg) {
+        File file = cfgPath.getParent().resolve(DATA_FILE).toFile();
+        if (!file.exists()) {
+            throw new DDPException("Data file is missing: " + file);
+        }
+        dataCfg = ConfigFactory.parseFile(file);
+        dataCfg = dataCfg.resolveWith(varsCfg);
+
+        if (!studyCfg.getString("study.guid").equals(STUDY_GUID)) {
+            throw new DDPException("This task is only for the " + STUDY_GUID + " study!");
+        }
+
+        this.varsCfg = varsCfg;
+        this.gson = GsonUtil.standardGson();
+    }
+
+    @Override
+    public void run(Handle handle) {
+        StudyDto studyDto = handle.attach(JdbiUmbrellaStudy.class).findByStudyGuid(STUDY_GUID);
+
+        Config blockCfg = dataCfg.getConfig("block");
+        validateBlockDef(blockCfg);
+
+        String activityCode = varsCfg.getString("id.act.baseline_symptom");
+        prependNewBlockInPlace(handle, studyDto.getId(), activityCode, blockCfg);
+
+        activityCode = varsCfg.getString("id.act.longitudinal");
+        prependNewBlockInPlace(handle, studyDto.getId(), activityCode, blockCfg);
+    }
+
+    private void validateBlockDef(Config blockCfg) {
+        var validator = new GsonPojoValidator();
+        var def = gson.fromJson(ConfigUtil.toJson(blockCfg), ContentBlockDef.class);
+        List<JsonValidationError> errors = validator.validateAsJson(def);
+        if (!errors.isEmpty()) {
+            String msg = errors.stream()
+                    .map(JsonValidationError::toDisplayMessage)
+                    .collect(Collectors.joining(", "));
+            throw new DDPException("Block definition has validation errors: " + msg);
+        }
+    }
+
+    private void prependNewBlockInPlace(Handle handle, long studyId, String activityCode, Config blockCfg) {
+        long activityId = ActivityBuilder.findActivityId(handle, studyId, activityCode);
+        LOG.info("Prepending block for activity {}...", activityCode);
+
+        ActivityVersionDto versionDto = handle.attach(JdbiActivityVersion.class)
+                .getActiveVersion(activityId)
+                .orElseThrow(() -> new DDPException("Could not find latest version for activity " + activityCode));
+
+        long timestamp = versionDto.getRevStart();
+        List<Long> sectionIds = handle.attach(JdbiFormActivityFormSection.class)
+                .findOrderedSectionIdsByActivityIdAndTimestamp(activityId, timestamp);
+        if (sectionIds.isEmpty()) {
+            throw new DDPException("Could not find sections for activity " + activityCode);
+        }
+
+        long sectionId = sectionIds.get(0);
+        long revisionId = versionDto.getRevId();
+        int displayOrder = SectionBlockDao.DISPLAY_ORDER_GAP / 2; // Current first block starts at this gap, so use something smaller.
```

Comment:
Can I ask how you find this anchor?  Is there a snippet of SQL to share?

Reply:
`SectionBlockDao` will insert the first block with `DISPLAY_ORDER_GAP`. This is seen in the code in `SectionBlockDao L166`. This was meant to leave "room" for us to prepend stuff (which we're leveraging now!). We can also confirm manually by querying the `form_activity__form_section` table for the section, then the `form_section__block` table for the listing of blocks.